### ISSUE TITLE
Serialization of nested generic object throws Type variable 'XY' can not be resolved

### ISSUE THREAD

**Issue by janpachol** [FasterXML/jackson-databind#537]
When jackson deserializes object of class which has nested generic type ```<E extends AnnotatedValue<ID>, ID>```, it throws

```
Type variable 'ID' can not be resolved (with context of class core.object.StaticCodebookTest$CbFailing)
com.fasterxml.jackson.databind.JsonMappingException
	at com.fasterxml.jackson.databind.SerializerProvider._createAndCacheUntypedSerializer(SerializerProvider.java:1042)
	at com.fasterxml.jackson.databind.SerializerProvider.findValueSerializer(SerializerProvider.java:445)
	at com.fasterxml.jackson.databind.SerializerProvider.findTypedValueSerializer(SerializerProvider.java:599)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:92)
	at com.fasterxml.jackson.databind.ObjectMapper._configAndWriteValue(ObjectMapper.java:2866)
	at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:2323)
	at core.object.StaticCodebookTest.test(StaticCodebookTest.java:29)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:45)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:15)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:42)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:263)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:68)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:47)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:231)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:60)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:50)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:222)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:300)
	at org.apache.maven.surefire.junit4.JUnit4TestSet.execute(JUnit4TestSet.java:53)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:123)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:104)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:164)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:110)
	at org.apache.maven.surefire.booter.SurefireStarter.invokeProvider(SurefireStarter.java:175)
	at org.apache.maven.surefire.booter.SurefireStarter.runSuitesInProcessWhenForked(SurefireStarter.java:107)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:68)
Caused by: java.lang.IllegalArgumentException: Type variable 'ID' can not be resolved (with context of class core.object.StaticCodebookTest$CbFailing)
	at com.fasterxml.jackson.databind.type.TypeBindings.findType(TypeBindings.java:170)
	at com.fasterxml.jackson.databind.type.TypeFactory._fromVariable(TypeFactory.java:847)
	at com.fasterxml.jackson.databind.type.TypeFactory._constructType(TypeFactory.java:400)
	at com.fasterxml.jackson.databind.type.TypeFactory._fromParamType(TypeFactory.java:801)
	at com.fasterxml.jackson.databind.type.TypeFactory._constructType(TypeFactory.java:391)
	at com.fasterxml.jackson.databind.type.TypeBindings._resolveBindings(TypeBindings.java:305)
	at com.fasterxml.jackson.databind.type.TypeBindings._resolve(TypeBindings.java:203)
	at com.fasterxml.jackson.databind.type.TypeBindings.findType(TypeBindings.java:121)
	at com.fasterxml.jackson.databind.type.TypeFactory._fromVariable(TypeFactory.java:847)
	at com.fasterxml.jackson.databind.type.TypeFactory._constructType(TypeFactory.java:400)
	at com.fasterxml.jackson.databind.type.TypeBindings.resolveType(TypeBindings.java:102)
	at com.fasterxml.jackson.databind.introspect.AnnotatedWithParams.getType(AnnotatedWithParams.java:87)
	at com.fasterxml.jackson.databind.introspect.AnnotatedMethod.getType(AnnotatedMethod.java:110)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory._constructWriter(BeanSerializerFactory.java:703)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.findBeanProperties(BeanSerializerFactory.java:557)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.constructBeanSerializer(BeanSerializerFactory.java:344)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.findBeanSerializer(BeanSerializerFactory.java:263)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory._createSerializer2(BeanSerializerFactory.java:222)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.createSerializer(BeanSerializerFactory.java:152)
	at com.fasterxml.jackson.databind.SerializerProvider._createUntypedSerializer(SerializerProvider.java:1077)
	at com.fasterxml.jackson.databind.SerializerProvider._createAndCacheUntypedSerializer(SerializerProvider.java:1037)
	... 35 more
```

I've create test for reproduction. Class ```CbFailing<E extends AnnotatedValue<ID>, ID>``` throws an exception, ```CbPassing<E extends AnnotatedValue, ID>``` is passing.

```java
    @Test
    public void test() throws JsonProcessingException {
        AnnotatedValueSimple<Integer> item = new AnnotatedValueSimple<Integer>(5);
        CbFailing<AnnotatedValueSimple<Integer>, Integer> codebook = new CbFailing<AnnotatedValueSimple<Integer>, Integer>(item);
        String json = new ObjectMapper().writeValueAsString(codebook);
    }
    
    private static class CbFailing<E extends AnnotatedValue<ID>, ID> {

        private E item;
        
        public CbFailing(E item) {
            this.item = item;
        }
        
        public E getItem() {
            return item;
        }
        
        public ID getId() {
            return item.getValue();
        }

    }

    private static class CbPassing<E extends AnnotatedValue, ID> {
        
        private E item;
        
        public CbPassing(E item) {
            this.item = item;
        }
        
        public E getItem() {
            return item;
        }
        
        public Object getId() {
            return item.getValue();
        }
        
    }
```

**Comment by cowtowncoder** [comment]
I think this may now be resolved with Jackson 2.7 (under `master` branch). Will try to test it.

**Comment by cowtowncoder** [comment]
Passes for 2.7.0; assuming fixed.
